qmi/core/transport.py

```diff
@@ -1071,29 +1092,47 @@ def read_until(self, message_terminator: bytes, timeout: Optional[float]) -> byt
             else:
                 raise QMI_InstrumentException("Error attempting to read bytes until {!r} char received".format(
                     message_terminator)) from err
+
         finally:
             # Revert terminator and timeout to previous value
             self._safe_instr.term_char = old_term_char
             self._safe_instr.timeout = old_timeout
 
-        return data
+        ret = self._read_buffer
+        self._read_buffer = bytes()
+        return ret
+
+    def read_until_timeout(self, nbytes: int, timeout: float) -> bytes:
+        try:
+            return self.read(nbytes, timeout)
+
+        except QMI_TimeoutException:
+            # Return whatever was read until timeout and clear the buffer.
+            ret = self._read_buffer
+            self._read_buffer = bytes()
+            return ret
 
     def discard_read(self) -> None:
         self._check_is_open()
         old_timeout = self._safe_instr.timeout
-        self._safe_instr.timeout = 0.0
+        self._safe_instr.timeout = 0.0  # Immediate read
+        # Clear any possible data in read buffer first
+        self._read_buffer = bytes()
         while True:
             try:
-                _ = self._safe_instr.read_raw(1)
+                self._safe_instr.read_raw(1)
+
             except vxi11.vxi11.Vxi11Exception as err:
                 if err.err == 15:
                     break
                 else:
+                    self._safe_instr.timeout = old_timeout
                     raise QMI_InstrumentException("Error attempting to read a byte") from err
+
             except TimeoutError:
                 break
-            finally:
```

Comment:
minor: why did you remove this finally? seems like a better way to make sure the old timeout is restored?

Reply:
Because the try-except is in a `while` loop and doesn't each `try` finish with `finally`, resetting the timeout after the first  loop?